### ISSUE TITLE
AssertionError: not all umis are the same length(!):  4 - 5

### ISSUE THREAD

**Issue by monalisa6hota** [CGATOxford/UMI-tools#461]
Hello,
 i am trying to run UMI-tools to remove duplicated reads based on UMI. But getting error "AssertionError: not all umis are the same length(!):  4 - 5". Can anyone suggest how to resolve this error?

### I am using following command :
umi_tools dedup --stdin=filtered.bam --umi-separator=":" --log=LOGFILE  --output-stats=stats.txt  -S output.bam > OUTFILE

### This is the head of my bam file:

A00609:116:H7JCGDSXY:1:2171:13937:29825 16      1       10534   255     3S96M2D21M31S   *       0       0       CGCAGTACCACCGAAATCTGTGCAGAGGAGAACGCAGCTCCGCCCTCGCGGTGCTCTCCGGGTCTGTGCTGAGGAGAACGCAACTCCGCCGTCGCAAAGGCGCCGCGCCGGCGCAGACGCCCCCATGTACTCTGCGTTGATACCACTGCTT FFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFF,FFFFFFFFF,FFFFFFFFFFFFFFFFFF:FFFF:FF:FFFFFFF:FFFFFFF:FFFFF:FFFF  NH:i:1  HI:i:1  AS:i:103        nM:i:3  RE:A:I  xf:i:0  li:i:0  CR:Z:CCGTGAGAGAACGCGT   CY:Z::FF:F:FF:FFF,F:F   CB:Z:CCGTGAGAGAACGCGT-1 UR:Z:TTTTCCGCACTT       UY:ZG:Z:cells:0:1:H7JCGDSXY:1

### I am getting following  error: 

Traceback (most recent call last):
  File "/home/ubuntu/miniconda3/bin/umi_tools", line 11, in <module>
    sys.exit(main())
  File "/home/ubuntu/miniconda3/lib/python3.8/site-packages/umi_tools/umi_tools.py", line 61, in main
    module.main(sys.argv)
  File "/home/ubuntu/miniconda3/lib/python3.8/site-packages/umi_tools/dedup.py", line 329, in main
    reads, umis, umi_counts = processor(
  File "/home/ubuntu/miniconda3/lib/python3.8/site-packages/umi_tools/network.py", line 419, in __call__
    clusters = self.UMIClusterer(counts, threshold)
  File "/home/ubuntu/miniconda3/lib/python3.8/site-packages/umi_tools/network.py", line 367, in __call__
    assert max(len_umis) == min(len_umis), (
AssertionError: not all umis are the same length(!):  4 - 5

**Comment by TomSmithCGAT** [comment]
By default, `umi_tools dedup` expects the UMI +/- cell barcode to be encoded in the read name. E.g, in the read above, it will take `29825` to be the read name. You can ignore the error since the issue is defining where the UMI is encoded in the command.

It looks like you have output from cellranger, with UMI in bam tag `UR` and cell barcode in bam tags `CR` and `CB`. From the [cell ranger documentation](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/bam), we can see that `CR` is the cell barcode directly from the fastq sequence, and `CB` is post error correction, in which case, you'll want to use `CB` I assume.

If you add the following to your command, `umi_tools` will deduplicate the reads in each cell independently, which I assume is what you want to achieve. You can drop the `--cell-tag` if not.

`--extract-umi-method=tag --umi-tag=UR --cell-tag=CB`

Also note that running `-output-stats` can slow down the processing and increase memory usage considerably. If you find this is the case, you may want to collect stats on a suitable subset of the data.

**Comment by monalisa6hota** [comment]
Thank you so much for your quick response. I tried with these options and the dedup ran fine.

**Comment by iammrtza** [comment]
Hi dear,

I am getting the same error saying

Traceback (most recent call last):
    File "/Users/morteza/opt/anaconda3/bin/umi_tools", line 11, in <module>
    sys.exit(main())
File "/Users/morteza/opt/anaconda3/lib/python3.8/site-packages/umi_tools/umi_tools.py", line 61, in main
module.main(sys.argv)
File "/Users/morteza/opt/anaconda3/lib/python3.8/site-packages/umi_tools/dedup.py", line 329, in main
reads, umis, umi_counts = processor(
    File "/Users/morteza/opt/anaconda3/lib/python3.8/site-packages/umi_tools/network.py", line 419, in __call__
    clusters = self.UMIClusterer(counts, threshold)
    File "/Users/morteza/opt/anaconda3/lib/python3.8/site-packages/umi_tools/network.py", line 367, in __call__
    assert max(len_umis) == min(len_umis), (
        AssertionError: not all umis are the same length(!):  4 - 12

my bam file looks like this (mapping is done by bowtie1):
seq32063_TCCCCGCCC	0	hsa-let-7a-5p	1	255	22M	*	0	0	TGAGGTAGTAGGTTGTATAGTT	IIIIIIIIIIIIIIIIIIIIII	XA:i:0	MD:Z:22	NM:i:0
seq34741_GGCTTAAGCAC	0	hsa-let-7a-5p	1	255	20M	*	0	0	TGAGGTAGTAGGTTGTATAG	IIIIIIIIIIIIIIIIIIII	XA:i:0	MD:Z:20	NM:i:0
seq37476_ACGCCTCCGCCC	0	hsa-let-7a-5p	1	255	5M	*	0	0	TGAGG	IIIII	XA:i:0	MD:Z:5	NM:i:0

Could you please help me with this? Thank you.

**Comment by TomSmithCGAT** [comment]
Hi @iammrtza. In the 3 lines above, the UMIs are all different lengths, so the error message appears to be correct. Overall, it appears your UMI lengths are between 4-12. They need to all be the same length.

What was the command you used to extract the UMIs from the fastq?

**Comment by iammrtza** [comment]
Hi Tom,

My Illumina reads structure is depicted [here](https://resources.qiagenbioinformatics.com/manuals/biomedicalgenomicsanalysis/current/index.php?manual=Create_UMI_Reads_miRNA.html). After removing adapter/junk (using cutadapt), I did grep for the "common sequence" and what is left was UMI and I added those to the reads headers in FASTA file. Then I removed the "common sequence" using cutadapt again.
